rfc/system/5741-binary-idl-with-message-pack.md

```diff
@@ -0,0 +1,596 @@
+```
+
+#### JSON Schema
+The JSON Schema of `BaseModel` will be generated by Pydantic's API.
+```python
+@dataclass
+class DC_inside:
+    a: int
+    b: float
+
+@dataclass
+class DC:
+    a: int
+    b: float
+    c: str
+    d: Dict[str, int]
+    e: DC_inside
+
+class MyDCModel(BaseModel):
+    dc: DC
+
+my_dc = MyDCModel(dc=DC(a=1, b=2.0, c="3", d={"4": 5}, e=DC_inside(a=6, b=7.0)))
+my_dc.model_json_schema()
+"""
+{'$defs': {'DC': {'properties': {'a': {'title': 'A', 'type': 'integer'}, 'b': {'title': 'B', 'type': 'number'}, 'c': {'title': 'C', 'type': 'string'}, 'd': {'additionalProperties': {'type': 'integer'}, 'title': 'D', 'type': 'object'}, 'e': {'$ref': '#/$defs/DC_inside'}}, 'required': ['a', 'b', 'c', 'd', 'e'], 'title': 'DC', 'type': 'object'}, 'DC_inside': {'properties': {'a': {'title': 'A', 'type': 'integer'}, 'b': {'title': 'B', 'type': 'number'}}, 'required': ['a', 'b'], 'title': 'DC_inside', 'type': 'object'}}, 'properties': {'dc': {'$ref': '#/$defs/DC'}}, 'required': ['dc'], 'title': 'MyDCModel', 'type': 'object'}
+"""
+```
+
+### FlyteCtl
+
+In FlyteCtl, we can construct input for the execution.
+
+When we receive `SimpleType.STRUCT`, we can construct a `Binary IDL Object` using the following logic in `flyteidl/clients/go/coreutils/literals.go`:
+
+```go
+if newT.Simple == core.SimpleType_STRUCT {
+    if _, isValueStringType := v.(string); !isValueStringType {
+        byteValue, err := msgpack.Marshal(v)
+        if err != nil {
+            return nil, fmt.Errorf("unable to marshal to json string for struct value %v", v)
+        }
+        strValue = string(byteValue)
+    }
+}
+```
+
+This is how users can create an execution by using a YAML file:
+```bash
+flytectl create execution --execFile ./flytectl/create_dataclass_task.yaml -p flytesnacks -d development
+```
+
+Example YAML file (`create_dataclass_task.yaml`):
+```yaml
+iamRoleARN: ""
+inputs:
+  input:
+    a: 1
+    b: 3.14
+    c: example_string
+    d:
+      "1": 100
+      "2": 200
+    e:
+      a: 1
+      b: 3.14
+envs: {}
+kubeServiceAcct: ""
+targetDomain: ""
+targetProject: ""
+task: dataclass_example.dataclass_task
+version: OSyTikiBTAkjBgrL5JVOVw
+```
+
+### FlyteCopilot
+
+When we need to pass an attribute access value to a copilot task, we must modify the code to convert a Binary Literal value with the `msgpack` tag into a primitive value.
+
+(Currently, we will only support primitive values.)
+
+You can reference the relevant section of code here:
+
+[FlyteCopilot - Data Download](https://github.com/flyteorg/flyte/blob/7989209e15600b56fcf0f4c4a7c9af7bfeab6f3e/flytecopilot/data/download.go#L88-L95)
+
+### FlyteConsole
+#### How users input into launch form?
+When FlyteConsole receives a literal type of `SimpleType.STRUCT`, the input method depends on the availability of a JSON schema:
+
+1. No JSON Schema provided:
+
+Input is expected as an `Object` (e.g., `{"a": 1}`).
```

Comment:
This is a Javascript Object right?  Let's add that.

Reply:
Can you also add a section about pasting?  How do users paste in something?  Can add to unresolved questions if you want.  If this is a Javascript object, we should allow json pasting, but what about yaml?  What about msgpack bytes if they were copied from a binary file?